src/term.rs

```diff
@@ -562,6 +565,55 @@ impl Term {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct SharedTerm {
+    term: Rc<Term>,
+}
+
+impl SharedTerm {
+    pub fn new(term: Term) -> Self {
+        Self {
+            term: Rc::new(term),
+        }
+    }
+
+    pub fn as_value(self) -> Term {
+        Rc::try_unwrap(self.term).unwrap_or_else(|rc| Term::clone(&rc))
+    }
+}
+
+impl AsRef<Term> for SharedTerm {
+    fn as_ref(&self) -> &Term {
+        self.term.as_ref()
+    }
+}
+
+impl AsMut<Term> for SharedTerm {
+    fn as_mut(&mut self) -> &mut Term {
+        Rc::make_mut(&mut self.term)
+    }
+}
+
+impl From<SharedTerm> for Term {
+    fn from(st: SharedTerm) -> Self {
+        st.as_value()
+    }
+}
+
+impl Deref for SharedTerm {
+    type Target = Term;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl DerefMut for SharedTerm {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.as_mut()
```

Comment:
same, good remark

Reply:
Funny thing, there is nothing in the doc requiring this trait to have a cheap implementation.
I would be tempted to keep this implementation in this case: 
- most of the times, this deref will be cheap,
- it makes the code less bloated than having to use a dedicated method all the times.

Reply:
I respectfully disagree. I don't know in what context someone would interpret "dereferencing a mutable smart pointer" as potentially deep-cloning arbitrary-sized vectors. `Rc` and `Cow` are two examples in the stdlib that are such cloning-on-mut smart pointers, and neither implement `DerefMut`. Because we are implementing a wrapper over `Rc`, doing as `Rc` does unless we have a good reason not to should be a good guideline.

Also, I found saving a few chars writing `*foo` instead of `foo.mk_mut()` to be a weak motivation. The latter makes things clearer. Also, if I remove the `DerefMut` implementation, the code still compiles, so it seems it's just dead code currently.